test/cpp/jit/test_lite_interpreter.cpp

```diff
@@ -606,11 +608,102 @@ TEST(LiteInterpreterTest, DuplicatedClassTypeModuleInfo) {
     }
   }
 
-  // The current approach is not able to distinguish between A0 and A1,
-  // which have the same class type. Hence, it only records module
-  // information for A1.
+  // class A(nn.Module):
+  //   def __init__(self):
+  //     super(A, self).__init__()
+
+  //   def forward(self, x):
+  //     return x + 1
+
+  // class B(nn.Module):
+  //   def __init__(self):
+  //     super(B, self).__init__()
+  //     self.A0 = A()
+  //     self.A1 = A()
+
+  //   def forward(self, x):
+  //     return self.A0(x) + self.A1(x)
+
+  // There are 3 module information strings here.
+  // "top(B).forward": for the add operator in top.
+  // "top(B).A0(B).forward": for the add operator in A0.
+  // "top(B).A1(A).forward": for the add operator in A1.
+
+  std::unordered_set<std::string> expected_result(
+      {"top(B).forward", "top(B).A0(A).forward", "top(B).A1(A).forward"});
+  AT_ASSERT(module_debug_info_set == expected_result);
+}
+
+TEST(LiteInterpreterTest, NestedCallModuleInfoCase) {
+  Module a("A");
+  a.define(R"JIT(
+    def forward(self, x):
+      return x + 1
+  )JIT");
+  Module b("B");
+  b.define(R"JIT(
+    def forward(self, x):
+      return x + 2
+  )JIT");
+  Module c("C");
+  c.register_module("A0", a);
+  c.register_module("B0", b);
+  c.define(R"JIT(
+    def forward(self, x):
+      return self.A0.forward(self.B0.forward(x)) + 1
+  )JIT");
+
+  std::stringstream ss;
+  c._save_for_mobile(ss, {}, true);
+  mobile::Module bc = _load_for_mobile(ss);
+
+  std::unordered_set<std::string> module_debug_info_set;
+  size_t pc = 0;
+  while (true) {
+    try {
+      std::string module_info = bc.get_forward_method_debug_info(pc);
+      if (!module_info.empty() && module_info != "<no module info>") {
+        module_debug_info_set.insert(module_info);
+      }
+      ++pc;
+    } catch (const std::exception& e) {
+      break;
+    }
+  }
+
+  // class A(nn.Module):
+  //   def __init__(self):
+  //     super(A, self).__init__()
+
+  //   def forward(self, x):
+  //     return x + 1
+
+  // class B(nn.Module):
+  //   def __init__(self):
+  //     super(B, self).__init__()
+
+  //   def forward(self, x):
+  //     return x + 2
+
+  // class C(nn.Module):
+  //   def __init__(self):
+  //     super(B, self).__init__()
+  //     self.A0 = A()
+  //     self.B0 = B()
+
+  //   def forward(self, x):
+  //     return self.A0.forward(self.B0.forward(x)) + 1
+
+  // There are 3 module information strings here.
+  // "top(C).forward": for the add operator in top.
+  // "top(C).A0(A).forward": for the add operator in A0 (second top).
+  // "top(C).A0(A).forward.B0(B).forward": for the add operator in B0 (third
+  // top).
+
   std::unordered_set<std::string> expected_result(
-      {"top(B).forward", "top(B).A1(A).forward"});
+      {"top(C).forward",
+       "top(C).A0(A).forward",
+       "top(C).A0(A).forward.B0(B).forward"});
```

Comment:
I think this should be 2 calls:

"top(C).A0(A).forward",
"top(C).B0(B).forward",

i.e. there's no nesting, since the code

```
//   def forward(self, x):
//     return self.A0.forward(self.B0.forward(x)) + 1
```

is equivalent to the code:

```
//   def forward(self, x):
//     t0 = self.B0.forward(x)
//     return self.A0.forward(t0) + 1
```

Reply:
Yes that's right. There might be something wrong with `Inline(*graph)`, or even before that. I'm investigating the root cause.

Reply:
This behavior is correct, we originally have 3 graphs (~functions):
```
graph_A(%x):
  %r = aten::add(%x, 1)
  return %r
graph_B(%x):
  %r = aten::add(%x, 2)
  return %r
graph_C(%input):
  %t1 = Call[graph='B'](%x) # syntax is made up, in reality we would have prim::CallFunctions here
  %t2 = Call[graph='A'](%t1)
  return %r
```

After inlining (which is replacing calls with the corresponding graph's body):
```
graph_C(%input):
  %t1 = aten::add(%input, 2) # inlined from graph B, callstack: ["B::forward"]
  %t2 = aten::add(%t1, 1) # inlined from graph A, callstack: ["A::forward"]
  return %r
```

There is no call of graph B in graph A and vice versa.

Reply:
I tried to dump the graph before inline and after inline. Is the following expected?
`
before inline: graph(%self.1 : __torch__.C,
      %x.1 : Tensor):
  %2 : __torch__.A = prim::GetAttr[name="A0"](%self.1)
  %3 : __torch__.B = prim::GetAttr[name="B0"](%self.1)
  %4 : Tensor = prim::CallMethod[name="forward"](%3, %x.1) # <string>:3:29
  %5 : Tensor = prim::CallMethod[name="forward"](%2, %4) # <string>:3:13
  return (%5)

before inline: graph(%self : __torch__.B,
      %x.1 : Tensor):
  %2 : int = prim::Constant[value=1]()
  %3 : int = prim::Constant[value=2]() # <string>:3:17
  %4 : Tensor = aten::add(%x.1, %3, %2) # <string>:3:13
  return (%4)

after inline: graph(%self : __torch__.B,
      %x.1 : Tensor):
  %2 : int = prim::Constant[value=1]()
  %3 : int = prim::Constant[value=2]() # <string>:3:17
  %4 : Tensor = aten::add(%x.1, %3, %2) # <string>:3:13
  return (%4)

before inline: graph(%self : __torch__.A,
      %x.1 : Tensor):
  %2 : int = prim::Constant[value=1]() # <string>:3:17
  %3 : Tensor = aten::add(%x.1, %2, %2) # <string>:3:13
  return (%3)

after inline: graph(%self : __torch__.A,
      %x.1 : Tensor):
  %2 : int = prim::Constant[value=1]() # <string>:3:17
  %3 : Tensor = aten::add(%x.1, %2, %2) # <string>:3:13
  return (%3)

after inline: graph(%self.1 : __torch__.C,
      %x.1 : Tensor):
  %2 : __torch__.A = prim::GetAttr[name="A0"](%self.1)
  %3 : __torch__.B = prim::GetAttr[name="B0"](%self.1)
  %6 : int = prim::Constant[value=2]() # <string>:3:17
  %7 : int = prim::Constant[value=1]()
  %8 : Tensor = aten::add(%x.1, %6, %7) # <string>:3:13
  %9 : int = prim::Constant[value=1]() # <string>:3:17
  %10 : Tensor = aten::add(%8, %9, %9) # <string>:3:13
  return (%10)
`

Reply:
Yes, that's exactly what is expected to happen: graphs for A and B remained unchanged (they contained no calls, thus nothing was inlined into them) and in the graph for C nodes `prim::CallMethod` were replaced with bodies of graphs A and B (the body is simply an `aten::add` and a `prim::Constant`).

Reply:
@ZolotukhinM unrelated to the discussion here, but when I look at

```
after inline: graph(%self.1 : __torch__.C,
      %x.1 : Tensor):
  %2 : __torch__.A = prim::GetAttr[name="A0"](%self.1)
  %3 : __torch__.B = prim::GetAttr[name="B0"](%self.1)
  %6 : int = prim::Constant[value=2]() # <string>:3:17
  %7 : int = prim::Constant[value=1]()
  %8 : Tensor = aten::add(%x.1, %6, %7) # <string>:3:13
  %9 : int = prim::Constant[value=1]() # <string>:3:17
  %10 : Tensor = aten::add(%8, %9, %9) # <string>:3:13
  return (%10)
```

I see `%2` and `%3`, which are essentially unused later. I'm wondering if these will be eliminated later or if the side effects of fetching these attributes using `prim::GetAttr` are significant?

Reply:
Yes, they will be cleaned up by the dead code elimination pass. Also, two identical constants (%7 and %9) will be merged into one by the constant pooling pass.